### PR TITLE
[#215] Suggest existing poetry venvs that are in existing modules

### LINKED ISSUE #215: Suggest existing poetry venvs that are in existing modules
**Is your feature request related to a problem? Please describe.**
When adding a Poetry environment, sometimes the venv already exists and I just want to register it. But the panel only shows the root directory's environment if it exists and is not yet mapped - I would like this to work when there are multiple Poetry projects in the same repo.

**Describe the solution you'd like**
I'd like the dropdown to include all the environments located in existing modules that have not yet been associated.

**Describe alternatives you've considered**
Associating by adding new also works, but it's a lot slower and shows all modules instead of just the ones that don't have SDKs yet. I would like to sweep through and just add ones that I don't have associations for.

**Additional context**

### PR DESCRIPTION
**Pull request by roxchkplusony**
Fixes #215

### CODE REVIEW

**Comment by roxchkplusony** [comment]
@koxudaxi I must admit that I'm a little uncomfortable with how to test the plugin well. I did test manually in IJ Ultimate. There is some change in behavior here: module path is used instead of project path if it's non-null now.

File: src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt [hunk 1]
```diff
@@ -55,20 +58,32 @@ class PyAddExistingPoetryEnvPanel(private val project: Project?,
                 .panel
         add(formPanel, BorderLayout.NORTH)
         addInterpretersAsync(sdkComboBox) {
-            detectPoetryEnvs(module, existingSdks, context, project?.basePath ?: newProjectPath)
+            val existingSdkPaths = sdkHomes(existingSdks)
+            val moduleSdks = allModules(project).parallelStream().flatMap { module ->
```
Comment (roxchkplusony) [inline src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt:62]: Does not take a crazy long time to load if there are many modules (I have 45):
Add_Python_Interpreter_and_Project_Structure

File: src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt [hunk 2]
```diff
@@ -55,20 +58,32 @@ class PyAddExistingPoetryEnvPanel(private val project: Project?,
...
+                val sdks = detectPoetryEnvs(module, existingSdkPaths, module.basePath)
+                        .filterNot { it.isAssociatedWithAnotherModule(module) }
+                sdks.forEach { sdkToModule.putIfAbsent(it.name, module) }
+                sdks.stream()
+            }.toList()
+            val rootSdks = detectPoetryEnvs(module, existingSdkPaths, project?.basePath ?: newProjectPath)
                     .filterNot { it.isAssociatedWithAnotherModule(module) }
+            val moduleSdkPaths = moduleSdks.map { it.name }.toSet()
+            val sdks = rootSdks.filterNot { moduleSdkPaths.contains(it.name) } + moduleSdks
+            sdks.sortedBy { it.name }
```
Comment (roxchkplusony) [inline src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt:72]: Sorted alphabetically, which is understandable and searchable (alternative was by module name):
Add_Python_Interpreter_and_Project_Structure

File: src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt [hunk 3]
```diff
@@ -55,20 +58,32 @@ class PyAddExistingPoetryEnvPanel(private val project: Project?,
...
+            val moduleSdks = allModules(project).parallelStream().flatMap { module ->
+                val sdks = detectPoetryEnvs(module, existingSdkPaths, module.basePath)
+                        .filterNot { it.isAssociatedWithAnotherModule(module) }
+                sdks.forEach { sdkToModule.putIfAbsent(it.name, module) }
+                sdks.stream()
+            }.toList()
+            val rootSdks = detectPoetryEnvs(module, existingSdkPaths, project?.basePath ?: newProjectPath)
                     .filterNot { it.isAssociatedWithAnotherModule(module) }
+            val moduleSdkPaths = moduleSdks.map { it.name }.toSet()
+            val sdks = rootSdks.filterNot { moduleSdkPaths.contains(it.name) } + moduleSdks
```
Comment (roxchkplusony) [inline src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt:71]: Ensure we only have one entry for the project if there's a module at the same level:

Comment (roxchkplusony) [reply src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt:71]: Ensure that associated sdks are not in the dropdown:
Project_Structure_and_data-mesh
Add_Python_Interpreter_and_Project_Structure_and_data-mesh

File: src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt [hunk 4]
```diff
@@ -55,20 +58,32 @@ class PyAddExistingPoetryEnvPanel(private val project: Project?,
...
     }
 
     override fun validateAll(): List<ValidationInfo> = listOfNotNull(validateSdkComboBox(sdkComboBox, this))
 
     override fun getOrCreateSdk(): Sdk? {
         return when (val sdk = sdkComboBox.selectedSdk) {
-            is PyDetectedSdk ->
-                setupPoetrySdkUnderProgress(project, module, existingSdks, newProjectPath,
+            is PyDetectedSdk -> {
+                val mappedModule = sdkToModule[sdk.name] ?: module
```
Comment (roxchkplusony) [inline src/com/koxudaxi/poetry/PyAddExistingPoetryEnvPanel.kt:81]: Created sdk matches the corresponding module's name:
Project_Structure_and_data-mesh

**Comment by koxudaxi** [comment]
I can't get the existing SDK 🤔

**Comment by koxudaxi** [comment]
This problem happened after just creating a new project with a poetry environment.

**Comment by koxudaxi** [comment]
@roxchkplusony 
I tested patterns in PyCharm.
It looks good 😄 
I have pushed it as `0.11.3`.

Thank you very much.

**Comment by roxchkplusony** [comment]
I'm so glad, thank you for your leadership!

**Pull request merged**